admin/admin_server.cc

```diff
@@ -0,0 +1,337 @@
+    }
+  }
+
+  res.setStatus(HTTPResponse::HTTP_OK);
+  res.setContentType("application/json");
+  res.send() << R"({"message": "common tags updated"})";
+}
+
+Poco::SharedPtr<Object> fmt_meter_object(spectator::Meter* meter) {
+  Object::Ptr obj = new Object(true);
+  obj->set("name", fmt::format("{}", meter->MeterId().Name()));
+
+  Object::Ptr tag_obj = new Object(true);
+  for (auto tag_it : meter->MeterId().GetTags()) {
+    tag_obj->set(fmt::format("{}", tag_it.key), fmt::format("{}", tag_it.value));
+  }
+  obj->set("tags", tag_obj);
+
+  return obj;
+}
+
+void GET_metrics(HTTPServerRequest& req, HTTPServerResponse& res, spectator::Registry& registry) {
+  Object::Ptr obj = new Object(true);
+
+  Poco::JSON::Array::Ptr age_gauges = new Poco::JSON::Array(true);
+  for (auto it : registry.AgeGauges()) {
+    auto meter = fmt_meter_object((spectator::Meter*)it);
+    meter->set("value", fmt::format("{}", it->Value()));
+    age_gauges->add(meter);
+  }
+  obj->set("age_gauges", age_gauges);
+
+  Poco::JSON::Array::Ptr counters = new Poco::JSON::Array(true);
+  for (auto it : registry.Counters()) {
+    auto meter = fmt_meter_object((spectator::Meter*)it);
+    meter->set("value", fmt::format("{}", it->Count()));
+    counters->add(meter);
+  }
+  obj->set("counters", counters);
+
+  Poco::JSON::Array::Ptr dist_summaries = new Poco::JSON::Array(true);
+  for (auto it : registry.DistSummaries()) {
+    auto meter = fmt_meter_object((spectator::Meter*)it);
+    meter->set("value", fmt::format("{}", it->TotalAmount()));
+    dist_summaries->add(meter);
+  }
+  obj->set("dist_summaries", dist_summaries);
+
+  Poco::JSON::Array::Ptr gauges = new Poco::JSON::Array(true);
+  for (auto it : registry.Gauges()) {
+    auto meter = fmt_meter_object((spectator::Meter*)it);
+    meter->set("value", fmt::format("{}", it->Get()));
+    gauges->add(meter);
+  }
+  obj->set("gauges", gauges);
+
+  Poco::JSON::Array::Ptr max_gauges = new Poco::JSON::Array(true);
+  for (auto it : registry.MaxGauges()) {
+    auto meter = fmt_meter_object((spectator::Meter*)it);
+    meter->set("value", fmt::format("{}", it->Get()));
+    max_gauges->add(meter);
+  }
+  obj->set("max_gauges", max_gauges);
+
+  Poco::JSON::Array::Ptr mono_counters = new Poco::JSON::Array(true);
+  for (auto it : registry.MonotonicCounters()) {
+    auto meter = fmt_meter_object((spectator::Meter*)it);
+    meter->set("value", fmt::format("{}", it->Delta()));
+    mono_counters->add(meter);
+  }
+  obj->set("mono_counters", mono_counters);
+
+  Poco::JSON::Array::Ptr timers = new Poco::JSON::Array(true);
+  for (auto it : registry.Timers()) {
+    auto meter = fmt_meter_object((spectator::Meter*)it);
+    meter->set("value", fmt::format("{}", it->TotalTime()));
+    timers->add(meter);
+  }
+  obj->set("timers", timers);
+
+  res.setStatus(HTTPResponse::HTTP_OK);
+  res.setContentType("application/json");
+  obj->stringify(res.send());
+}
+
+spectator::Id parse_id(const std::string& id_str) {
```

Comment:
Is this the same logic for parsing ids on the incoming data updates? Could that logic be reused so we ensure they stay consistent?

Reply:
It is close, but not the same (`spectatord::get_measurement`). Due to the need to (1) extract data values and (2) fail poorly-formatted protocol messages in `get_measurement`, there is not a clean way to re-use the id-parsign logic without adding additional conditional branching to the whole thing.

This variation of id parsing logic is a bit looser - it will always produce an Id, even with a poorly formatted string - it just won't match anything in the Registry. Given how this stuff will be used, it does not feel like we need to be any more strict with it.